fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java

```diff
@@ -149,15 +150,19 @@ public Optional<QueryPredicate> toQueryPredicate(@Nonnull final CorrelationIdent
                 // it seems this is a constant expression, try to evaluate it.
                 return tryBoxSelfAsConstantPredicate();
             }
-        } else if (childrenCount == 2) {
+        }
+        Verify.verify(childrenCount == 2);
+        final var leftChild = it.next();
+        final var rightChild = it.next();
+        if (comparisonType == Comparisons.Type.IN) {
+            final var values = Objects.requireNonNull(((LiteralValue<?>) rightChild).getLiteralValue());
```

Comment:
We miss an important chance of optimisation here.

Say we have the following `in` predicate: `(3+4) in (1, 1+1, 3, 5, 5+2)`, we could pre-evaluate this to `true` which could cascade other execution optimisations.

E.g. if we can prove that the predicate is always `false` we can prune an entire execution.

Reply:
That's right! while we cannot evaluate it completely in the compile-time (because we do not have a physical implementation of IN function) we can actually as an optimization evaluate the left and right sides of it. so that the example you demonstrated reduces to `7 in (1, 2, 3, 5, 7)`. Adding that with other changes.